<center> <img src = https://raw.githubusercontent.com/AndreyRysistov/DatasetsForPandas/main/hh%20label.jpg alt="drawing" style="width:400px;">

# <center> Проект: Анализ вакансий из HeadHunter
   

In [84]:
import pandas as pd
import psycopg2
import requests
from bs4 import BeautifulSoup

In [97]:
DBNAME = 'project_sql'
USER = 'skillfactory'
PASSWORD = 'cCkxxLVrDE8EbvjueeMedPKt'
HOST = '84.201.134.129'
PORT = 5432

In [98]:
connection = psycopg2.connect(
    dbname=DBNAME,
    user=USER,
    host=HOST,
    password=PASSWORD,
    port=PORT
)

# Юнит 3. Предварительный анализ данных

1. Напишите запрос, который посчитает количество вакансий в нашей базе (вакансии находятся в таблице vacancies).

In [ ]:
query = f'''
SELECT COUNT(*) vacancies_count
FROM vacancies
'''

In [49]:
df = pd.read_sql_query(query, connection)
display(df)

C:\Users\Максим\AppData\Local\Temp\ipykernel_132\3475273297.py:1: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query, connection)


,vacancies_count
0,49197


2. Напишите запрос, который посчитает количество работодателей (таблица employers).

In [ ]:
query = f'''
SELECT COUNT(*) employers_count
FROM employers
'''

In [8]:
df = pd.read_sql_query(query, connection)
display(df)

C:\Users\Максим\AppData\Local\Temp\ipykernel_132\3475273297.py:1: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query, connection)


,employers_count
0,23501


3. Посчитате с помощью запроса количество регионов (таблица areas).

In [ ]:
query = f'''
SELECT COUNT(*) areas_count
FROM areas
'''

In [10]:
df = pd.read_sql_query(query, connection)
display(df)

C:\Users\Максим\AppData\Local\Temp\ipykernel_132\3475273297.py:1: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query, connection)


,areas_count
0,1362


4. Посчитате с помощью запроса количество сфер деятельности в базе (таблица industries).

In [ ]:
query = f'''
SELECT COUNT(*) industries_count
FROM industries
'''

In [12]:
df = pd.read_sql_query(query, connection)
display(df)

C:\Users\Максим\AppData\Local\Temp\ipykernel_132\3475273297.py:1: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query, connection)


,industries_count
0,294


***

In [ ]:
# выводы по предварительному анализу данных

В базе имеются 49197 вакансий, 23501 работодателей, 1362 региона и 294 сфер деятельности.

# Юнит 4. Детальный анализ вакансий

1. Напишите запрос, который позволит узнать, сколько (cnt) вакансий в каждом регионе (area).
Отсортируйте по количеству вакансий в порядке убывания.

In [ ]:
query = f'''
SELECT a.name area, COUNT(v.id) cnt
FROM vacancies v
JOIN areas a on v.area_id = a.id
GROUP BY a.name
ORDER BY cnt DESC
LIMIT 5
'''

In [14]:
df = pd.read_sql_query(query, connection)
display(df)

C:\Users\Максим\AppData\Local\Temp\ipykernel_132\3475273297.py:1: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query, connection)


,area,cnt
0,Москва,5333
1,Санкт-Петербург,2851
2,Минск,2112
3,Новосибирск,2006
4,Алматы,1892


2. Напишите запрос, чтобы определить у какого количества вакансий заполнено хотя бы одно из двух полей с зарплатой.

In [ ]:
query = '''
SELECT COUNT(*) vacansies
FROM vacancies
WHERE salary_from IS NOT NULL OR salary_to IS NOT NULL
'''

In [16]:
df = pd.read_sql_query(query, connection)
display(df)

C:\Users\Максим\AppData\Local\Temp\ipykernel_132\3475273297.py:1: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query, connection)


,vacansies
0,24073


3. Найдите средние значения для нижней и верхней границы зарплатной вилки. Округлите значения до целого.

In [ ]:
query = '''
SELECT ROUND(AVG(salary_from)) avg_salary_from, ROUND(AVG(salary_to)) avg_salary_to
FROM vacancies
'''

In [18]:
df = pd.read_sql_query(query, connection)
display(df)

C:\Users\Максим\AppData\Local\Temp\ipykernel_132\3475273297.py:1: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query, connection)


,avg_salary_from,avg_salary_to
0,71065.0,110537.0


4. Напишите запрос, который выведет количество вакансий для каждого сочетания типа рабочего графика (schedule) и типа трудоустройства (employment), используемого в вакансиях. Результат отсортируйте по убыванию количества.


In [ ]:
query = '''
SELECT schedule, employment, COUNT(*) vacancies_count
FROM vacancies
GROUP BY schedule, employment
ORDER BY vacancies_count DESC
'''

In [20]:
df = pd.read_sql_query(query, connection)
display(df)

C:\Users\Максим\AppData\Local\Temp\ipykernel_132\3475273297.py:1: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query, connection)


,schedule,employment,vacancies_count
0,Полный день,Полная занятость,35367
1,Удаленная работа,Полная занятость,7802
2,Гибкий график,Полная занятость,1593
3,Удаленная работа,Частичная занятость,1312
4,Сменный график,Полная занятость,940
5,Полный день,Стажировка,569
6,Вахтовый метод,Полная занятость,367
7,Полный день,Частичная занятость,347
8,Гибкий график,Частичная занятость,312
9,Полный день,Проектная работа,141


5. Напишите запрос, выводящий значения поля Требуемый опыт работы (experience) в порядке возрастания количества вакансий, в которых указан данный вариант опыта.

In [21]:
query = '''
SELECT experience, COUNT(*) vacancies_count
FROM vacancies
GROUP BY experience
ORDER BY vacancies_count
'''

In [22]:
df = pd.read_sql_query(query, connection)
display(df)

C:\Users\Максим\AppData\Local\Temp\ipykernel_132\3475273297.py:1: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query, connection)


,experience,vacancies_count
0,Более 6 лет,1337
1,Нет опыта,7197
2,От 3 до 6 лет,14511
3,От 1 года до 3 лет,26152


***

In [ ]:
# выводы по детальному анализу вакансий

1.   Пятерка лидеров по количеству вакансий: Москва, Санкт-Петербург, Минск, Новосибирск, Алматы.
2.   Количества вакансий где заполнено хотя бы одно из двух полей с зарплатой - 24073.
3.   Нижняя и верхняя границы зарплатной вилки - от 71065 до 110537.

# Юнит 5. Анализ работодателей

1. Напишите запрос, который позволит узнать, какие работодатели находятся на первом и пятом месте по количеству вакансий.

In [ ]:
query = '''
SELECT e.name AS employer, COUNT(v.id) AS cnt
FROM vacancies v
JOIN employers e ON v.employer_id = e.id
GROUP BY e.name
ORDER BY cnt DESC
LIMIT 5;
'''

In [24]:
df = pd.read_sql_query(query, connection)
display(df)

C:\Users\Максим\AppData\Local\Temp\ipykernel_132\3475273297.py:1: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query, connection)


,employer,cnt
0,Яндекс,1933
1,Ростелеком,491
2,Тинькофф,444
3,СБЕР,428
4,Газпром нефть,331


2. Напишите запрос, который для каждого региона выведет количество работодателей и вакансий в нём.
Среди регионов, в которых нет вакансий, найдите тот, в котором наибольшее количество работодателей.


In [ ]:
query_1 = '''
SELECT a.name area_name, count(DISTINCT e.id) employers_count, count(DISTINCT v.id) vacancies_count
FROM areas a
LEFT JOIN employers e ON e.area = a.id
LEFT JOIN vacancies v ON v.area_id = a.id
GROUP BY a.id
ORDER BY employers_count DESC;
'''

query_2 = '''
SELECT a.name area_name, count(DISTINCT e.id) employers_count, count(DISTINCT v.id) vacancies_count
FROM areas a
LEFT JOIN employers e ON e.area = a.id
LEFT JOIN vacancies v ON v.area_id = a.id
GROUP BY a.id
HAVING count(DISTINCT v.id) = 0
ORDER BY employers_count DESC
LIMIT 1;
'''

In [30]:
df = pd.read_sql_query(query_1, connection)
display(df)

df = pd.read_sql_query(query_2, connection)
display(df)

C:\Users\Максим\AppData\Local\Temp\ipykernel_132\994754631.py:1: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_1, connection)


,area_name,employers_count,vacancies_count
0,Москва,5864,5333
1,Санкт-Петербург,2217,2851
2,Минск,1115,2112
3,Алматы,721,1892
4,Екатеринбург,609,1698
...,...,...,...
1357,Янино-1,0,0
1358,Адыге-Хабль,0,1
1359,Эркен-Шахар,0,0
1360,Индустриальный (Краснодарский край),0,0


C:\Users\Максим\AppData\Local\Temp\ipykernel_132\994754631.py:4: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_2, connection)


,area_name,employers_count,vacancies_count
0,Россия,410,0


3. Для каждого работодателя посчитайте количество регионов, в которых он публикует свои вакансии. Отсортируйте результат по убыванию количества.


In [ ]:
query = '''
SELECT e.name, count(DISTINCT v.area_id) AS vacancies_count
FROM vacancies v
JOIN employers e ON v.employer_id=e.id
GROUP BY e.name
ORDER BY vacancies_count DESC;
'''

In [92]:
df = pd.read_sql_query(query, connection)
display(df)

/tmp/ipykernel_11127/3475273297.py:1: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query, connection)


,name,vacancies_count
0,Яндекс,181
1,Ростелеком,152
2,Спецремонт,116
3,Поляков Денис Иванович,88
4,ООО ЕФИН,71
...,...,...
14761,UniSol,1
14762,UNISTORY LLC,1
14763,UNIT6,1
14764,United Distribution,1


4. Напишите запрос для подсчёта количества работодателей, у которых не указана сфера деятельности.

In [ ]:
query = '''
SELECT COUNT(employers_with_industry.id) AS employers_without_industry
FROM
  (SELECT e.id, e.name, ei.industry_id
  FROM employers e
  LEFT JOIN employers_industries ei ON e.id=ei.employer_id) AS employers_with_industry
WHERE employers_with_industry.industry_id IS NULL;
'''

In [94]:
df = pd.read_sql_query(query, connection)
display(df)

/tmp/ipykernel_11127/3475273297.py:1: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query, connection)


,employers_without_industry
0,8419


5. Напишите запрос, чтобы узнать название компании, находящейся на третьем месте в алфавитном списке (по названию) компаний, у которых указано четыре сферы деятельности.

In [ ]:
query = '''
SELECT e.name
FROM employers e
LEFT JOIN employers_industries ei ON ei.employer_id = e.id
GROUP BY e.name
HAVING COUNT(ei.industry_id) = 4
ORDER BY e.name
OFFSET 2
LIMIT 1
'''

In [100]:
df = pd.read_sql_query(query, connection)
display(df)

/tmp/ipykernel_11127/3475273297.py:1: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query, connection)


,name
0,2ГИС


6. С помощью запроса выясните, у какого количества работодателей в качестве сферы деятельности указана Разработка программного обеспечения.


In [ ]:
query = '''
SELECT count(e.id) employers_count
FROM employers_industries ei
JOIN employers e ON ei.employer_id = e.id
JOIN industries i ON i.id = ei.industry_id
WHERE i.name = 'Разработка программного обеспечения'
'''

In [102]:
df = pd.read_sql_query(query, connection)
display(df)

/tmp/ipykernel_11127/3475273297.py:1: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query, connection)


,employers_count
0,3553


7. Для компании «Яндекс» выведите список регионов-миллионников, в которых представлены вакансии компании, вместе с количеством вакансий в этих регионах. Также добавьте строку Total с общим количеством вакансий компании. Результат отсортируйте по возрастанию количества.

Список городов-милионников надо взять [отсюда](https://ru.wikipedia.org/wiki/%D0%93%D0%BE%D1%80%D0%BE%D0%B4%D0%B0-%D0%BC%D0%B8%D0%BB%D0%BB%D0%B8%D0%BE%D0%BD%D0%B5%D1%80%D1%8B_%D0%A0%D0%BE%D1%81%D1%81%D0%B8%D0%B8).

Если возникнут трудности с этим задание посмотрите материалы модуля  PYTHON-17. Как получать данные из веб-источников и API.

In [119]:
# код для получения списка городов-милионников

url = "https://ru.wikipedia.org/wiki/Города-миллионеры_России"
headers = {'User-Agent': 'Python'}

response = requests.get(url, headers=headers)
bs = BeautifulSoup(response.text, 'html.parser')
big_cities_lst = pd.read_html(str(bs))[0]['Город'].tolist()
big_cities_str = ", ".join([f"'{city}'" for city in big_cities_lst])
print(cities_str)

'Москва', 'Санкт-Петербург', 'Новосибирск', 'Екатеринбург', 'Казань', 'Красноярск', 'Нижний Новгород', 'Челябинск', 'Уфа', 'Краснодар', 'Самара', 'Ростов-на-Дону', 'Омск', 'Воронеж', 'Пермь', 'Волгоград'


/tmp/ipykernel_11127/3733535005.py:8: FutureWarning: Passing literal html to 'read_html' is deprecated and will be removed in a future version. To read from a literal string, wrap it in a 'StringIO' object.
  big_cities_lst = pd.read_html(str(bs))[0]['Город'].tolist()


In [ ]:
# текст запроса
query = f'''
SELECT a.name AS city, COUNT(v.id) AS vacancies
FROM areas a
JOIN vacancies v ON a.id = v.area_id
JOIN employers e ON v.employer_id = e.id
WHERE e.name = 'Яндекс' AND a.name IN ({cities_str})
GROUP BY a.name
ORDER BY vacancies
'''

In [124]:
# результат запроса
df = pd.read_sql_query(query, connection)

total_vacancies = df["vacancies"].sum()
df.loc[len(df)] = ["Total", total_vacancies]
display(df)

/tmp/ipykernel_11127/4001442375.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query, connection)


,city,vacancies
0,Омск,21
1,Челябинск,22
2,Красноярск,23
3,Волгоград,24
4,Пермь,25
5,Казань,25
6,Ростов-на-Дону,25
7,Уфа,26
8,Самара,26
9,Краснодар,30


***

In [ ]:
# выводы по анализу работодателей

На первом месте по количеству вакансий находится Яндекс, а на пятом месте Газпром нефть
Среди регионов, в которых нет вакансий, найдите тот, в котором наибольшее количество работодателей - это Россия
Максимальное количество регионов, в которых работодатели публикуют свои вакансии - 181
Количество работодателей, у которых не указана сфера деятельности - 8419
У какого количества работодателей в качестве сферы деятельности указана «Разработка программного обеспечения» - 3553

# Юнит 6. Предметный анализ

1. Сколько вакансий имеет отношение к данным?

Считаем, что вакансия имеет отношение к данным, если в её названии содержатся слова 'data' или 'данн'.

*Подсказка: Обратите внимание, что названия вакансий могут быть написаны в любом регистре.*


In [127]:
query = '''
SELECT COUNT(*) AS data_related_vacancies
FROM public.vacancies
WHERE LOWER(name) LIKE '%data%' OR LOWER(name) LIKE '%данн%';
'''

In [128]:
df = pd.read_sql_query(query, connection)
display(df)

C:\Users\Максим\AppData\Local\Temp\ipykernel_132\3475273297.py:1: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query, connection)


,data_related_vacancies
0,1771


2. Сколько есть подходящих вакансий для начинающего дата-сайентиста?
Будем считать вакансиями для дата-сайентистов такие, в названии которых есть хотя бы одно из следующих сочетаний:
* 'data scientist'
* 'data science'
* 'исследователь данных'
* 'ML' (здесь не нужно брать вакансии по HTML)
* 'machine learning'
* 'машинн%обучен%'

** В следующих заданиях мы продолжим работать с вакансиями по этому условию.*

Считаем вакансиями для специалистов уровня Junior следующие:
* в названии есть слово 'junior' *или*
* требуемый опыт — Нет опыта *или*
* тип трудоустройства — Стажировка.


In [129]:
query = '''
SELECT COUNT(*) AS vacancies_count
FROM vacancies
WHERE (LOWER(name) LIKE '%data%scientist%' 
    OR LOWER(name) LIKE '%data%science%'
    OR LOWER(name) LIKE '%исследователь данных%'
    OR (LOWER(name) LIKE '%ml%' AND LOWER(name) NOT LIKE '%html%')
    OR LOWER(name) LIKE '%machine learning%'
    OR LOWER(name) LIKE '%машинн%обучен%')
    AND (LOWER(name) LIKE '%junior%'
    OR experience = 'Нет опыта'
    OR employment = 'Стажировка');
'''

In [130]:
df = pd.read_sql_query(query, connection)
display(df)

C:\Users\Максим\AppData\Local\Temp\ipykernel_132\3475273297.py:1: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query, connection)


,vacancies_count
0,51


3. Сколько есть вакансий для DS, в которых в качестве ключевого навыка указан SQL или postgres?

** Критерии для отнесения вакансии к DS указаны в предыдущем задании.*

In [131]:
query = '''
SELECT COUNT(*) AS vacancies_count
FROM public.vacancies
WHERE (LOWER(name) LIKE '%data%scientist%' 
    OR LOWER(name) LIKE '%data%science%'
    OR LOWER(name) LIKE '%исследователь данных%'
    OR (LOWER(name) LIKE '%ml%' AND LOWER(name) NOT LIKE '%html%')
    OR LOWER(name) LIKE '%machine learning%'
    OR LOWER(name) LIKE '%машинн%обучен%')
    AND (LOWER(key_skills) LIKE '%sql%' 
    OR LOWER(key_skills) LIKE '%postgres%');
'''

In [37]:
df = pd.read_sql_query(query, connection)
display(df)

C:\Users\Максим\AppData\Local\Temp\ipykernel_132\3475273297.py:1: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query, connection)


,vacancies_count
0,229


4. Проверьте, насколько популярен Python в требованиях работодателей к DS.Для этого вычислите количество вакансий, в которых в качестве ключевого навыка указан Python.

** Это можно сделать помощью запроса, аналогичного предыдущему.*

In [132]:
query = '''
SELECT COUNT(*) AS vacancies_count
FROM public.vacancies
WHERE (LOWER(name) LIKE '%data%scientist%' 
    OR LOWER(name) LIKE '%data%science%'
    OR LOWER(name) LIKE '%исследователь данных%'
    OR (LOWER(name) LIKE '%ml%' AND LOWER(name) NOT LIKE '%html%')
    OR LOWER(name) LIKE '%machine learning%'
    OR LOWER(name) LIKE '%машинн%обучен%')
    AND LOWER(key_skills) LIKE '%python%';
'''

In [133]:
df = pd.read_sql_query(query, connection)
display(df)

C:\Users\Максим\AppData\Local\Temp\ipykernel_132\3475273297.py:1: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query, connection)


,vacancies_count
0,357


5. Сколько ключевых навыков в среднем указывают в вакансиях для DS?
Ответ округлите до двух знаков после точки-разделителя.

In [136]:
# Интересно что в предыдущих заданиях верный ответ получался при использовании 
# (LOWER(name) LIKE '%ml%' AND LOWER(name) NOT LIKE '%html%')
# А в данном случае только при использовании (name LIKE '%ML%' AND name NOT LIKE '%HTML%')
query = '''
SELECT AVG(cardinality(string_to_array(key_skills, CHR(9)))) AS avg_skills_count
FROM vacancies
WHERE (LOWER(name) LIKE '%data scientist%' 
    OR LOWER(name) LIKE '%data science%'
    OR LOWER(name) LIKE '%исследователь данных%'
    OR (name LIKE '%ML%' AND name NOT LIKE '%HTML%')
    OR LOWER(name) LIKE '%machine learning%'
    OR LOWER(name) LIKE '%машинн%обучен%')
    
'''

In [137]:
df = pd.read_sql_query(query, connection)
display(df)

C:\Users\Максим\AppData\Local\Temp\ipykernel_132\3475273297.py:1: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query, connection)


,avg_skills_count
0,6.406032


6. Напишите запрос, позволяющий вычислить, какую зарплату для DS в **среднем** указывают для каждого типа требуемого опыта (уникальное значение из поля *experience*).

При решении задачи примите во внимание следующее:
1. Рассматриваем только вакансии, у которых заполнено хотя бы одно из двух полей с зарплатой.
2. Если заполнены оба поля с зарплатой, то считаем зарплату по каждой вакансии как сумму двух полей, делённую на 2. Если заполнено только одно из полей, то его и считаем зарплатой по вакансии.
3. Если в расчётах участвует null, в результате он тоже даст null (посмотрите, что возвращает запрос select 1 + null). Чтобы избежать этой ситуацию, мы воспользуемся функцией [coalesce](https://postgrespro.ru/docs/postgresql/9.5/functions-conditional#functions-coalesce-nvl-ifnull), которая заменит null на значение, которое мы передадим. Например, посмотрите, что возвращает запрос `select 1 + coalesce(null, 0)`

Выясните, на какую зарплату в среднем может рассчитывать дата-сайентист с опытом работы от 3 до 6 лет. Результат округлите до целого числа.

In [138]:
query = '''
SELECT experience, ROUND(AVG(COALESCE(((salary_from + salary_to)/2), salary_from, salary_to)))
FROM vacancies
WHERE (lower(vacancies.name) LIKE '%data scientist%' 
    OR lower(vacancies.name) LIKE '%data science%'
    OR lower(vacancies.name) LIKE '%исследователь данных%'
    OR (vacancies.name LIKE '%ML%' AND vacancies.name NOT LIKE '%HTML%')
    OR lower(vacancies.name) LIKE '%machine learning%'
    OR lower(vacancies.name) LIKE '%data science%'
    OR lower(vacancies.name) LIKE '%машинн%обучен%')
    AND
    (salary_from IS NOT NULL or salary_to IS NOT NULL)
GROUP BY vacancies.experience;
'''

In [139]:
df = pd.read_sql_query(query, connection)
display(df)

C:\Users\Максим\AppData\Local\Temp\ipykernel_132\3475273297.py:1: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query, connection)


,experience,round
0,Нет опыта,74643.0
1,От 1 года до 3 лет,139675.0
2,От 3 до 6 лет,243115.0


***

In [ ]:
# выводы по предметному анализу

Вакансий имеющих отношение к данным - 1771
Вакансий подходящих для начинающего дата сайентиста - 51
Вакансий для DS где в качестве ключевого навыка указан SQL или postgres - 229
Вакансий для DS где в качестве ключевого навыка указан python - 357
Ключевых навыков в среднем указывают в вакансиях для DS - 6,41
На какую зарплату в среднем может рассчитывать дата-сайентист с опытом работы от 3 до 6 лет - 243115

# Общий вывод по проекту

In [ ]:
# подведем итог исследования, обобщите выводы
# здесь можно (это будет плюсом) провести дополнительные исследования данных, сделать прогнозы, продумать варианты продолжения исследования